### ISSUE TITLE
No space left on device (not due to path preservation)

### ISSUE THREAD

**Issue by dma9** [trapexit/mergerfs#598]
General description
===================
Any write attempt on the mounted union fs returns a "no space left on device" message, even though there is significant free space available. For example, touch, mv or cp. Attempts to do so directly on the underlying drives do however work. In /etc/fstab, category.create has been set to lfs and moveonenospec=true. There are multiple drives in the union fs that have sufficient space for the write operation.

I suspect this is a result of error on my part, but I can't quite seem to figure out what I've done that would cause this to happen. The only thing of note that I had been doing before this problem seemed to crop up was to fix a container that had locked up because it ran out of storage space. It was used as a samba server for the mergerfs filesystem. I adjusted the disk allocation for the container and restarted it and it seemed to be find, but shortly after started getting the no space left on device messages.

Any suggestions you might have would be most appreciated.

Expected behavior
=================
The write operation would execute without returning the error message.

Actual behavior
===============
The write operation returns a "no space left on device" message and does not execute.

**Precise** steps to reproduce the behavior
===========================================

> Explicitly list **all** steps to reproduce. Preferably create a minimal example of the problem using standard command line tools. The more variables (apps, settings, etc.) that are involved the more difficult it is to debug. Also, **please** be sure to have read all of the README. It contains a lot of information regarding known system and user issues.

The simplest example would be touch. I change to the mountpoint of union fs and run "touch test.txt". This generates the following error message: "touch: cannot touch 'text.txt': No space left on device.

System information
==================
Please provide as much of the following information as possible:

* [ ] mergerfs version: `mergerfs -V`
```
mergerfs version: 2.25.1
FUSE library version: 2.9.8-mergerfs
fusermount version: 2.9.7
using FUSE kernel interface version 7.19

```
* [ ] mergerfs settings: `cat /etc/fstab` or the command line arguments
```
# <file system> <mount point> <type> <options> <dump> <pass>
/dev/zvol/rpool/swap none swap sw 0 0
proc /proc proc defaults 0 0

# SnapRAID Data Disks
/dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840HAZR-part1 /mnt/data/data1 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840HJCF-part1 /mnt/data/data2 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840HHXZ-part1 /mnt/data/data3 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z8412N4E-part1 /mnt/data/data4 ext4 defaults 0 2
# /dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840Z00H-part1 /mnt/data/data4 ext4 defaults 0 2
# /dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840H9R6-part1 /mnt/data/data4 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840HJL2-part1 /mnt/data/data5 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840HH30-part1 /mnt/data/data6 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840HB2W-part1 /mnt/data/data7 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840HHRB-part1 /mnt/data/data8 ext4 defaults 0 2

# SnapRAID Data Disks - additional disks added 2018-07-29
/dev/disk/by-id/ata-ST8000DM004-2CX188_WCT0F655-part1 /mnt/data/data9 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000DM004-2CX188_WG8029BK-part1 /mnt/data/data10 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000DM004-2CX188_WCT0F8JN-part1 /mnt/data/data11 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000DM004-2CX188_WCT0GC4X-part1 /mnt/data/data12 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000DM004-2CX188_WCT0EJPW-part1 /mnt/data/data13 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000DM004-2CX188_WCT0EXB9-part1 /mnt/data/data14 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000DM004-2CX188_WCT0FP2G-part1 /mnt/data/data15 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000DM004-2CX188_WCT0EWGT-part1 /mnt/data/data16 ext4 defaults 0 2

# SnapRAID Parity Disks
/dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840HAYB-part1 /mnt/parity/parity1 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840HJGL-part1 /mnt/parity/parity2 ext4 defaults 0 2

# SnapRAID Parity Disks - additional disks added 2018-07-29
/dev/disk/by-id/ata-ST8000DM004-2CX188_WCT0FQT6-part1 /mnt/parity/parity3 ext4 defaults 0 2
/dev/disk/by-id/ata-ST8000DM004-2CX188_WG80298K-part1 /mnt/parity/parity4 ext4 defaults 0 2

#MergerFS Pooling
/mnt/data/* /vdata fuse.mergerfs defaults,allow_other,direct_io,use_ino,category.create=lfs,moveonenospc=true,minfreespace=20G,fsname=mergerfsPool 0 0
```

* [ ] Linux version: `uname -a`
`Linux fava2 4.15.18-12-pve #1 SMP PVE 4.15.18-35 (Wed, 13 Mar 2019 08:24:42 +0100) x86_64 GNU/Linux
`
* [ ] Versions of any additional software being used
`snapraid v11.3 by Andrea Mazzoleni, http://www.snapraid.it`

* [ ] List of drives, filesystems, & sizes: `df -h`
```
Filesystem                    Size  Used Avail Use% Mounted on
udev                           63G     0   63G   0% /dev
tmpfs                          13G  9.9M   13G   1% /run
rpool/ROOT/pve-1              225G   86G  140G  38% /
tmpfs                          63G   43M   63G   1% /dev/shm
tmpfs                         5.0M  4.0K  5.0M   1% /run/lock
tmpfs                          63G     0   63G   0% /sys/fs/cgroup
mergerfsPool                  117T   87T   28T  76% /vdata
/dev/sda1                     7.3T  7.2T   12G 100% /mnt/data/data1
/dev/sde1                     7.3T  7.2T   20G 100% /mnt/data/data3
/dev/sdi1                     7.3T  7.2T   14G 100% /mnt/data/data5
/dev/sdn1                     7.3T  7.2T  6.1G 100% /mnt/data/data7
/dev/sdj1                     7.3T  7.2T   14G 100% /mnt/data/data6
/dev/sds1                     7.3T  7.2T  144G  99% /mnt/parity/parity1
/dev/sdd1                     7.3T  7.2T   19G 100% /mnt/data/data2
/dev/sdt1                     7.3T  7.2T  144G  99% /mnt/parity/parity2
/dev/sdo1                     7.3T  7.2T   20G 100% /mnt/data/data8
rpool                         140G     0  140G   0% /rpool
/dev/sdg1                     7.3T  6.0T  1.2T  84% /mnt/data/data12
/dev/sdq1                     7.3T  1.6T  5.7T  22% /mnt/data/data14
rpool/ROOT                    140G     0  140G   0% /rpool/ROOT
/dev/sdm1                     7.3T  7.2T   20G 100% /mnt/data/data9
/dev/sdf1                     7.3T  971G  6.2T  14% /mnt/data/data16
/dev/sdb1                     7.3T  724G  6.5T  10% /mnt/data/data15
/dev/sdk1                     7.3T  7.2T  143G  99% /mnt/parity/parity3
/dev/sdc1                     7.3T  5.6T  1.7T  78% /mnt/data/data11
rpool/data                    140G     0  140G   0% /rpool/data
/dev/sdp1                     7.3T  7.2T  143G  99% /mnt/parity/parity4
rpool/data/subvol-100-disk-1  224G  195G   30G  87% /rpool/data/subvol-100-disk-1
rpool/data/subvol-109-disk-1   16G  723M   16G   5% /rpool/data/subvol-109-disk-1
/dev/sdr1                     7.3T  7.2T   18G 100% /mnt/data/data10
/dev/sdh1                     7.3T  6.6T  563G  93% /mnt/data/data4
rpool/data/subvol-117-disk-0  4.0G  1.4G  2.7G  35% /rpool/data/subvol-117-disk-0
/dev/sdl1                     7.3T  904G  6.3T  13% /mnt/data/data13
rpool2                        1.3T  256K  1.3T   1% /rpool2
rpool2/subvol-100-disk-0      500G  281G  220G  57% /rpool2/subvol-100-disk-0
rpool2/subvol-101-disk-1      8.0G  817M  7.3G  10% /rpool2/subvol-101-disk-1
rpool2/subvol-103-disk-0      128G  1.2G  127G   1% /rpool2/subvol-103-disk-0
rpool2/subvol-104-disk-0       32G  7.2G   25G  23% /rpool2/subvol-104-disk-0
rpool2/subvol-105-disk-0       32G  4.2G   28G  14% /rpool2/subvol-105-disk-0
rpool2/subvol-106-disk-0      250G  540M  250G   1% /rpool2/subvol-106-disk-0
rpool2/subvol-107-disk-0       16G  1.1G   15G   7% /rpool2/subvol-107-disk-0
rpool2/subvol-108-disk-0       16G  940M   16G   6% /rpool2/subvol-108-disk-0
rpool2/subvol-109-disk-0       16G  911M   16G   6% /rpool2/subvol-109-disk-0
rpool2/subvol-110-disk-0       16G  3.5G   13G  22% /rpool2/subvol-110-disk-0
rpool2/subvol-111-disk-0       16G  1.6G   15G  10% /rpool2/subvol-111-disk-0
rpool2/subvol-112-disk-0       16G  871M   16G   6% /rpool2/subvol-112-disk-0
rpool2/subvol-113-disk-0      8.0G  859M  7.2G  11% /rpool2/subvol-113-disk-0
rpool2/subvol-114-disk-0      8.0G  879M  7.2G  11% /rpool2/subvol-114-disk-0
rpool2/subvol-115-disk-0      350G  133G  218G  38% /rpool2/subvol-115-disk-0
rpool2/subvol-116-disk-0      8.0G  4.1G  4.0G  51% /rpool2/subvol-116-disk-0
/dev/fuse                      30M   24K   30M   1% /etc/pve
tmpfs                          13G     0   13G   0% /run/user/0
```

* [ ] strace of application having problem: `strace -f -o /tmp/app.strace.txt <cmd>` or `strace -f -p <appPID> -o /tmp/app.strace.txt`
```
2605  execve("/usr/bin/touch", ["touch", "test.txt"], [/* 19 vars */]) = 0
2605  brk(NULL)                         = 0x55f29f0c8000
2605  access("/etc/ld.so.nohwcap", F_OK) = -1 ENOENT (No such file or directory)
2605  access("/etc/ld.so.preload", R_OK) = -1 ENOENT (No such file or directory)
2605  open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
2605  fstat(3, {st_mode=S_IFREG|0644, st_size=51104, ...}) = 0
2605  mmap(NULL, 51104, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f738be3c000
2605  close(3)                          = 0
2605  access("/etc/ld.so.nohwcap", F_OK) = -1 ENOENT (No such file or directory)
2605  open("/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
2605  read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0\4\2\0\0\0\0\0"..., 832) = 832
2605  fstat(3, {st_mode=S_IFREG|0755, st_size=1689360, ...}) = 0
2605  mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f738be3a000
2605  mmap(NULL, 3795296, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f738b887000
2605  mprotect(0x7f738ba1c000, 2097152, PROT_NONE) = 0
2605  mmap(0x7f738bc1c000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x195000) = 0x7f738bc1c000
2605  mmap(0x7f738bc22000, 14688, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f738bc22000
2605  close(3)                          = 0
2605  arch_prctl(ARCH_SET_FS, 0x7f738be3b4c0) = 0
2605  mprotect(0x7f738bc1c000, 16384, PROT_READ) = 0
2605  mprotect(0x55f29efe4000, 4096, PROT_READ) = 0
2605  mprotect(0x7f738be49000, 4096, PROT_READ) = 0
2605  munmap(0x7f738be3c000, 51104)     = 0
2605  brk(NULL)                         = 0x55f29f0c8000
2605  brk(0x55f29f0e9000)               = 0x55f29f0e9000
2605  open("/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
2605  fstat(3, {st_mode=S_IFREG|0644, st_size=1679776, ...}) = 0
2605  mmap(NULL, 1679776, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f738bc9f000
2605  close(3)                          = 0
2605  open("test.txt", O_WRONLY|O_CREAT|O_NOCTTY|O_NONBLOCK, 0666) = -1 ENOSPC (No space left on device)
2605  utimensat(AT_FDCWD, "test.txt", NULL, 0) = -1 ENOENT (No such file or directory)
2605  open("/usr/share/locale/locale.alias", O_RDONLY|O_CLOEXEC) = 3
2605  fstat(3, {st_mode=S_IFREG|0644, st_size=2995, ...}) = 0
2605  read(3, "# Locale name alias data base.\n#"..., 3072) = 2995
2605  read(3, "", 3072)                 = 0
2605  close(3)                          = 0
2605  open("/usr/share/locale/en_US.UTF-8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
2605  open("/usr/share/locale/en_US.utf8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
2605  open("/usr/share/locale/en_US/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
2605  open("/usr/share/locale/en.UTF-8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
2605  open("/usr/share/locale/en.utf8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
2605  open("/usr/share/locale/en/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
2605  write(2, "touch: ", 7)            = 7
2605  write(2, "cannot touch 'test.txt'", 23) = 23
2605  open("/usr/share/locale/en_US.UTF-8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
2605  open("/usr/share/locale/en_US.utf8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
2605  open("/usr/share/locale/en_US/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
2605  open("/usr/share/locale/en.UTF-8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
2605  open("/usr/share/locale/en.utf8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
2605  open("/usr/share/locale/en/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
2605  write(2, ": No space left on device", 25) = 25
2605  write(2, "\n", 1)                 = 1
2605  close(1)                          = 0
2605  close(2)                          = 0
2605  exit_group(1)                     = ?
2605  +++ exited with 1 +++
```

* [ ] strace of mergerfs while app tried to do it's thing: `strace -f -p <mergerfsPID> -o /tmp/mergerfs.strace.txt`
https://pastebin.com/8DhufrJX

**Comment by trapexit** [comment]
`2426  open("/mnt/data/data4/test.txt", O_WRONLY|O_CREAT|O_NONBLOCK|O_LARGEFILE, 0100666) = -1 ENOSPC (No space left on device)`

Are you using ext4 reserved space?

**Comment by dma9** [comment]
Ach. I'm terribly sorry for wasting your time. Your note triggered me to check something I should have checked at some point in trying to figure this out but for some reason didn't. Anyway, the problem appears to be that I've run out of inodes on /mnt/data/data4. 

Thanks very much for looking at this so quickly, and apologies once again for unnecessarily raising an issue.

On a somewhat related note, also wanted to take the opportunity to thank you for mergerfs. It's really great and very, very useful.

**Comment by trapexit** [comment]
No worries. I do mention running out of inodes in the docs but even right now I didn't think of it before considering it might be reserved space.

I could add checks for inode count but I suspect it'd result in just as many support questions but asking why the drive isn't being used. This more explicitly lets you know you're out. It's pretty uncommon regardless.

**Comment by dma9** [comment]
Could I perhaps trouble you on a somewhat related question? I've been ruminating on how to solve this problem with running out of inodes. As I have 20 drives, was hoping to avoid copying the data off each drive, reformatting to increase the number of inodes, then copying back. They're all formatted with ext4.  I looked into ZFS (which I understand doesn't have a fixed inode limit) but apparently there is no in-place conversion tool. Which led me to btrfs, which apparently does have an in-place conversion tool (and also no fixed inode limit). This seems to relatively expedient and would resolve my inode problem, but I was wondering if you had any thoughts regarding btrfs generally or using it in connection with mergerfs.

Just to be clear, the plan would be to convert each individual drive to btrfs while continuing to use mergerfs as the union fs. I would not be using any of the RAID functions of btrfs.

Any thoughts or suggestions you have would be most appreciated.

**Comment by trapexit** [comment]
I've tried btrfs in the past and had my drive become unrecoverably corrupted so my personal experience is not good with it but many use it just fine. I can't speak to the inplace change or fstransform which could do the same with other filessytems. 

I've moved to using XFS (the hard way, rsync'ing everything to other drives then reformatting and using mergerfs.balance) and have been happy with it but it doesn't (yet) have all the features of some modern ones.
